sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala

```diff
@@ -151,7 +152,15 @@ class TestHiveContext(sc: SparkContext) extends HiveContext(sc) {
 
   val describedTable = "DESCRIBE (\\w+)".r
 
+  val vs = new VariableSubstitution()
+
   protected[hive] class HiveQLQueryExecution(hql: String)
+    extends this.SubstitutedHiveQLQueryExecution(vs.substitute(hiveconf, hql))
+
+  // we should substitute variables in hql to pass the text to parseSql() as a parameter.
+  // Hive parser need substituted text. HiveContext.sql() does this but return a DataFrame,
+  // while we need a logicalPlan so we cannot reuse that.
+  protected[hive] class SubstitutedHiveQLQueryExecution(hql: String)
     extends this.QueryExecution(HiveQl.parseSql(hql)) {
     def hiveExec() = runSqlHive(hql)
```

Comment:
@adrian-wang how about adding the substitution in `HiveContext.runSqlHive` or `HiveContext.runHive`? Then we probably not necessary to change anything in `TestHive`.